File: nodejs/muse/algorithms/meta/SearchTest.js
'use strict'

const Search = require('muse/algorithms/meta/Search')
const SequenceBuilder = require('muse/util/SequenceBuilder')
const TestRunner = require('muse/util/TestRunner')

const testBinarySearch = () => {
  const size = 32768

  const arr = new Array(size)
  SequenceBuilder.packIncreasing(arr)

  if (Search.binarySearch(arr, -1) !== size) {
    return false
  }

  if (Search.binarySearch(arr, 2147483647) !== size) {
    return false
  }

  for (let i = 0; i < size; i++) {
    if (Search.binarySearch(arr, arr[i]) !== i) {
      return false
    }
  }

  return true
}

const testLinearSearch = () => {
  const size = 32768

  const arr = new Array(size)
  SequenceBuilder.packIncreasing(arr)

  if (Search.linearSearch(arr, -1) !== size) {
    return false
  }

  if (Search.linearSearch(arr, 2147483647) !== size) {
    return false
  }

  for (let i = 0; i < size; i++) {
    if (Search.linearSearch(arr, arr[i]) !== i) {
      return false
    }
  }

  return true
}

if (module === require.main) {
  TestRunner.pick(testBinarySearch)

  TestRunner.pick(testLinearSearch)
}

File: python/muse/algorithms/meta/SearchTest.py
from muse.algorithms.meta import Search
from muse.util import SequenceBuilder, TestRunner


def testBinarySearch() -> bool:
    size = 32768

    arr = [0] * size
    SequenceBuilder.packIncreasing(arr)

    if Search.binarySearch(arr, -1) != size:
        return False

    if Search.binarySearch(arr, 2_147_483_647) != size:
        return False

    for i, v in enumerate(arr):
        if Search.binarySearch(arr, v) != i:
            return False

    return True


def testLinearSearch() -> bool:
    size = 32768

    arr = [0] * size
    SequenceBuilder.packIncreasing(arr)

    if Search.linearSearch(arr, -1) != size:
        return False

    if Search.linearSearch(arr, 2_147_483_647) != size:
        return False

    for i, v in enumerate(arr):
        if Search.linearSearch(arr, v) != i:
            return False

    return True


def main() -> None:
    TestRunner.pick(testBinarySearch)

    TestRunner.pick(testLinearSearch)


if __name__ == "__main__":
    main()
